pyvrp/cpp/crossover/selective_route_exchange.cpp

```diff
@@ -8,6 +9,57 @@ using ClientSet = std::unordered_set<Client>;
 using Route = std::vector<Client>;
 using Routes = std::vector<Route>;
 
+namespace
+{
+double _computeRouteAngleCenter(ProblemData const &data, Route const &route)
+{
+    // Computes the route angle center. Assumes that the route is non-empty.
+    int cumulatedX = 0;
+    int cumulatedY = 0;
+
+    for (int node : route)
+    {
+        cumulatedX += data.client(node).x;
+        cumulatedY += data.client(node).y;
+    }
+
+    // This computes a pseudo-angle that sorts roughly equivalently to the atan2
+    // angle, but is much faster to compute. See the following post for details:
+    // https://stackoverflow.com/a/16561333/4316405.
+    auto const routeSize = static_cast<double>(route.size());
+    auto const dy = cumulatedY / routeSize - data.depot().y;
+    auto const dx = cumulatedX / routeSize - data.depot().x;
+    return std::copysign(1. - dx / (std::fabs(dx) + std::fabs(dy)), dy);
+}
+
+Routes _getNonEmptyRoutesSortedByAngleCenter(ProblemData const &data,
+                                             Routes const routes)
+{
+    std::vector<std::pair<double, int>> routePolarAngles;
+    routePolarAngles.reserve(data.numVehicles());
+
+    size_t numNonEmptyRoutes = 0;
+    for (size_t r = 0; r < data.numVehicles(); r++)
+        if (!routes[r].empty())
+        {
+            numNonEmptyRoutes++;
+            routePolarAngles.emplace_back(
+                _computeRouteAngleCenter(data, routes[r]), r);
+        }
+
+    // Sort angles of non-empty routes
+    std::sort(routePolarAngles.begin(),
+              routePolarAngles.begin() + numNonEmptyRoutes);
+
+    Routes resultRoutes;
+    resultRoutes.reserve(numNonEmptyRoutes);
+    for (size_t r = 0; r < numNonEmptyRoutes; r++)
+        resultRoutes.emplace_back(routes[routePolarAngles[r].second]);
+
+    return resultRoutes;
```

Comment:
I implemented your suggestion but note that it may affect performance as `angle` is now called `O(n log n)` times rather than `O(n)` times (where each call iterators over the complete route).